// @neftify/connect/dist/collection/components/connect-button/neftify-connect-button.js
// import {
//   Component,
//   Host,
//   h,
//   Element,
//   Fragment,
//   Event,
//   EventEmitter,
//   State,
//   Prop
// } from "@stencil/core";
// import {
//   text,
//   closeIcon,
//   neftifyIcon
// } from "./neftify-connect-button.templates";
// import Neftify from "@neftify/sdk";
// const neftify = new Neftify({ apiKey: "halla" });
// @Component({
//   tag: "neftify-connect-button",
//   styleUrl: "neftify-connect-button.scss",
//   shadow: true
// })
// export class NeftifyConnectButton {
//   // ========================== GLOBALS ==========================
//   // ------ ELEMENTS ------
//   /** Host element */
//   @Element() el: HTMLElement;
//   modalEl: HTMLNeftifyModalElement;
//   buttonEl: HTMLNeftifyButtonElement;
//   emailInputEl: HTMLNeftifyInputElement;
//   phoneInputEl: HTMLNeftifyInputElement;
//   codeInputEl: HTMLNeftifyInputElement;
//   // ------ STATES ------
//   @State() currentStep: number = 1;
//   @State() withPhone: boolean = true;
//   @State() hasConnected: boolean = false;
//   @State() email: string = "";
//   @State() phone: string | number = "";
//   @State() code: string | number = "";
//   @Prop({ mutable: true })
//   walletConnected: boolean = false;
//   // ------ EVENTS ------
//   /**
//    * Gets fired whenever the user has connected their wallet
//    */
//   @Event() private neftifyUserHasConnected: EventEmitter;
//   /**
//    * Gets fired whenever the user has disconnected their wallet
//    */
//   @Event() private neftifyUserHasDisconnected: EventEmitter;
//   // -------- HANDLERS --------
//   handleSignUpSignIn(withPhone) {
//     if (!withPhone) {
//       neftify
//         .createUserWithEmail(this.email)
//         .then(() => this.sendLoginCodeEmail(this.email))
//         .catch(error => console.error("⛔️ [Neftify ERROR]: ", error));
//     }
//   }
//   handleSubmitCode(withPhone) {
//     if (!withPhone) {
//       neftify
//         .loginWithEmailAndCode(this.email, this.code as string)
//         .then(response => {
//           this.hasConnected = true;
//           this.modalEl.hideModal();
//           this.neftifyUserHasConnected.emit(response);
//           console.log("[neftify user has connected]: ", response);
//         })
//         .catch(error => console.error("⛔️ [Neftify ERROR]: ", error));
//     }
//   }
//   // -------- GENERAL -------
//   sendLoginCodeEmail = email => {
//     return neftify
//       .sendLoginCode(email)
//       .then(() => (this.currentStep = 2))
//       .catch(error => console.log(error));
//   };
//   resetModalProgress() {
//     setTimeout(() => {
//       this.currentStep = 1;
//       this.withPhone = true;
//       this.hasConnected = false;
//       this.phoneInputEl ? (this.phoneInputEl.value = "") : null;
//       this.emailInputEl ? (this.emailInputEl.value = "") : null;
//       this.codeInputEl ? (this.codeInputEl.value = "") : null;
//     }, 250);
//   }
//   closeAndReset() {
//     this.modalEl.hideModal();
//     this.resetModalProgress();
//   }
//   // -------- RENDER --------
//   renderModalHeader = () => {
//     return (
//       <div class="neftify-connect-modal-header">
//         {neftifyIcon()}
//         <neftify-button onClick={() => this.closeAndReset()}>
//           {closeIcon()}
//         </neftify-button>
//       </div>
//     );
//   };
//   renderStepOneContent = () => {
//     return (
//       <Fragment>
//         <h2>
//           {text.general.title}
//         </h2>
//         <p>
//           {text.general.description}
//         </p>
//         <div class="first-step-input-wrapper">
//           {this.withPhone
//             ? <neftify-input
//                 key="phone-input"
//                 id="phone-input"
//                 ref={el => (this.phoneInputEl = el)}
//                 placeholder="123 456 7890"
//                 onNeftifyBlur={e => (this.phone = e.target.value)}
//                 type="text"
//                 maxlength={10}
//               />
//             : <neftify-input
//                 key="email-input"
//                 id="email-input"
//                 ref={el => (this.emailInputEl = el)}
//                 onNeftifyBlur={e => (this.email = e.target.value as string)}
//                 placeholder="youremail@email.com"
//                 type="text"
//               />}
//         </div>
//         <div class="neftify-connect-modal-content-switcher">
//           <div onClick={() => (this.withPhone = !this.withPhone)}>
//             Use {this.withPhone ? "email" : "phone"} instead
//           </div>
//         </div>
//       </Fragment>
//     );
//   };
//   renderStepTwoContent = () => {
//     return (
//       <Fragment>
//         <p>
//           {text.code.title} {this.withPhone ? "phone" : "email"}
//         </p>
//         <neftify-input
//           key="code-input"
//           onBlur={() => (this.codeInputEl.value = this.codeInputEl.value)}
//           ref={el => (this.codeInputEl = el)}
//           maxlength={6}
//           placeholder="000000"
//           type="text"
//         />
//         <div class="neftify-connect-modal-content-switcher">
//           <div>
//             {text.code.resend}
//           </div>
//         </div>
//       </Fragment>
//     );
//   };
//   renderModalActionButton = () => {
//     return (
//       <div class="neftify-connect-modal-action-wrapper">
//         {this.currentStep === 1
//           ? <neftify-button
//               slot="buttons"
//               onClick={() => this.handleSignUpSignIn(this.withPhone)}
//             >
//               {text.general.button}
//             </neftify-button>
//           : <neftify-button
//               slot="buttons"
//               onClick={() => this.handleSubmitCode(this.withPhone)}
//             >
//               {text.code.button}
//             </neftify-button>}
//       </div>
//     );
//   };
//   handleDisconnect = () => {
//     this.resetModalProgress();
//     this.walletConnected = false;
//     this.neftifyUserHasDisconnected.emit();
//   };
//   render() {
//     return (
//       <Host>
//         {!this.hasConnected && !this.walletConnected
//           ? <neftify-modal
//               class="neftify-connect-modal"
//               ref={el => (this.modalEl = el)}
//             >
//               {this.renderModalHeader()}
//               <div class="neftify-connect-modal-content">
//                 {this.currentStep === 1
//                   ? this.renderStepOneContent()
//                   : this.renderStepTwoContent()}
//               </div>
//               {this.renderModalActionButton()}
//             </neftify-modal>
//           : null}
//         {!this.hasConnected && !this.walletConnected
//           ? <neftify-button
//               class="neftify-connect-button"
//               ref={el => (this.buttonEl = el)}
//               onClick={() => this.modalEl.showModal()}
//             >
//               Connect wallet
//             </neftify-button>
//           : <neftify-button
//               class="neftify-connect-button"
//               onClick={() => {
//                 this.handleDisconnect();
//               }}
//             >
//               Disconnect
//             </neftify-button>}
//       </Host>
//     );
//   }
// }
